net/ipsec/ipsec.go

```diff
@@ -0,0 +1,395 @@
+// package IPsec provides primitives for establishing IPsec in the fastdp mode.
+package ipsec
+
+// TODO(mp) install iptables rule (raw OUTPUT) which drops marked non-ESP traffic.
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"sync"
+	"syscall"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/crypto/hkdf"
+
+	"github.com/weaveworks/mesh"
+)
+
+type SPI uint32
+
+const (
+	keySize = 36 // AES-GCM key 32 bytes + 4 bytes salt
+
+	mark    = uint32(0x1) << 17
+	markStr = "0x20000/0x20000"
+
+	table     = "mangle"
+	markChain = "WEAVE-IPSEC-MARK"
+	mainChain = "WEAVE-IPSEC"
+
+	mask   = (SPI(1) << (mesh.PeerShortIDBits)) - 1
+	spiMSB = SPI(1) << 31
+)
+
+// IPSec
+
+type IPSec struct {
+	sync.RWMutex
+	ipt *iptables.IPTables
+	rc  *connRefCount
+}
+
+func New() (*IPSec, error) {
+	ipt, err := iptables.New()
+	if err != nil {
+		return nil, errors.Wrap(err, "iptables new")
+	}
+
+	ipsec := &IPSec{
+		ipt: ipt,
+		rc:  newConnRefCount(),
+	}
+
+	return ipsec, nil
+}
+
+// Protect establishes IPsec between given peers.
+func (ipsec *IPSec) Protect(srcPeer, dstPeer mesh.PeerShortID, srcIP, dstIP net.IP, dstPort int, masterKey *[32]byte) (SPI, error) {
```

Comment:
Look into fixing #1737. Hopefully it's not that hard, and it's really quite an annoying bug anyway.